### ISSUE TITLE
semi-proposal: Is there a better way users could provide SSH/HTTPS/Git configuration?

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#1878]
Note: Once I feel a bit more confident about this or have more concrete thoughts, I will flesh out a real proposal on this following our new proposal process. Until then, this exists as a brain dump merely for gathering some small feedback and to remind myself of what I am thinking of exactly.

Sourcegraph occasionally requires SSH/HTTPS/Git configuration. See:

- https://docs.sourcegraph.com/admin/repo/auth#repositories-that-need-http-s-or-ssh-authentication
- https://github.com/sourcegraph/deploy-sourcegraph-docker#configuring-ssh-cloning
- https://github.com/sourcegraph/deploy-sourcegraph/blob/master/docs/configure.md#configure-repository-cloning-via-ssh

Problems with this current setup include:

1. We do not provide instructions to test the configuration in all docs (currently, only deploy-sourcegraph-docker does). The turnaround time between applying a change, testing that repository cloning works, etc. is painful and slow.
2. It is easy to mistake docs.sourcegraph.com as applying to Cluster setups, when in reality it only partially does.
3. It is easy to mess up the permissions of the SSH directory. For example, if your ssh directory is owned by root but Sourcegraph tries to clone as the containers `sourcegraph` user.
4. In cluster deployments, you must configure SSH on every gitserver instance uniformly. This is less of a problem in Kubernetes, but more of a problem in non-Kubernetes deployments.
5. In cluster deployments, you must store your secret SSH credentials elsewhere and be careful to not send PRs to us that contain them. We have to warn users about this, currently.
6. deploy-sourcegraph currently doesn't explain how to configure HTTPS authentication, but we do explain this for Server users. Do we need to for cluster users, too? https://docs.sourcegraph.com/admin/repo/auth#http-s-authentication-via-netrc

All of these issues (many of which could be addressed individually or through more documentation) makes me wonder: Is there a better way that we could have users provide SSH/HTTPS/Git configuration? Would it be valuable to invest time in doing so?

One possible proposal would be:

1. "To configure cloning of repositories, use the following: https://sourcegraph.example.com/site-admin/configure-cloning"
2. When you visit that page, an in-browser page / form allows you to specify:
    - `~/.ssh/id_rsa`
    - `~/.ssh/id_rsa.pem`
    - `~/.ssh/known_hosts`
    - `~/.ssh/config`
    - `~/.netrc`
    - `~/.gitconfig`
    - Additionally, it has an area where you can type in a `git clone <url>` and test cloning works.
3. Once you are finished, hit save and Sourcegraph will persist your changes to the database (site config) the same way it does with other secrets. (The changes are automatically propagated to all gitservers, you do not need to restart anything, do not need to mount any files, etc.)

Thoughts?

**Comment by tsenart** [comment]
I think the proper solution here would be to integrate with external secret managers like Vault. cc @beyang

**Comment by beyang** [comment]
Might be worth sending out a customer survey to understand how people do secret management inside their companies. We should integrate with whatever existing secret management a customer is already using for other internal apps/tools.

**Comment by tsenart** [comment]
I agree, but Vault is by far the most flexible as it can integrate in many different environments with its pluggable architecture.

Sent via Superhuman iOS ( https://sprh.mn/?vip=tomas@sourcegraph.com )

**Comment by keegancsmith** [comment]
docker and kubernetes have an existing secrets management solution. Vault is great, but I would try use a more generic abstraction (such as what k8s provides). I suppose moving from just reading secrets to writing them is quite a jump and has a lot of ground for us to cover to be generic. FYI I think vault is great, but targeting it would limit us since I would suspect most of our customers do not use it.

**Comment by tsenart** [comment]
> but targeting it would limit us since I would suspect most of our customers do not use it.

I am all for what integrates better in our customer's environments. It seems that we have a different understanding of how Vault can or can not do that smoothly. Could you elaborate on which limitations you foresee with it? I specifically was thinking of Vault *because* I think it'd offer us the most flexibility to integrate into heterogenous environments. 

- https://www.vaultproject.io/docs/secrets/index.html
- https://www.vaultproject.io/docs/auth/kubernetes.html
- https://www.vaultproject.io/docs/plugin/index.html

**Comment by tsenart** [comment]
Sounds good to me.

Sent via Superhuman iOS ( https://sprh.mn/?vip=tomas@sourcegraph.com )

**Comment by slimsag** [comment]
This issue is already backlogged btw.

**Comment by tsenart** [comment]
Related to #71

**Comment by keegancsmith** [comment]
keeping on backlog, out of scope for the work I wanted to tackle around external comms.
